### PR TITLE
Fix timeout for rd_kafka_query_watermark_offsets

### LINKED ISSUE #2588: rd_kafka_query_watermark_offsets API hang forever
Read the FAQ first: https://github.com/edenhill/librdkafka/wiki/FAQ

Description
===========
rd_kafka_query_watermark_offsets API will hang forever when the kafka cluster network encounter access restriction(network isolation)

How to reproduce
================
**I could reproduce this problem with latest librdkafka version**

1. launch 2 vm/docker instances(my local os is centos 6). A, B
2. install confluent-oss at instance A, start kafka with 3 broker services
     - brokerId: 1, port:9093
     - brokerId: 2, port:9094
     - brokerId: 3, port:9095
3. create a topic "test" for kafka with 3 partitions and replication-factor equal to 1, each broker should have a unique partition Id, assuming the "test" topic is with the following compositions:
     - brokerId: 1, port:9093 partitionId:0
     - brokerId: 2, port:9094 partitionId:1
     - brokerId: 3, port:9095 partitionId:2
4. at instance B, deploy the test program [main.go](https://github.com/edenhill/librdkafka/files/3770606/main.go.zip)
5. enable iptable service at instance A, just reject instance B's accessing for port 9095
6. Now run test program at instance A(test API QueryWatermarkOffsets), and it will hang(**the partitionId 2's broker is alive but is not accessible for instanceA**)
    - ./kafkatest -broker=$instanceB_IP:9093 -newAPI=true -topic=test -partitionId=2 -timeout=2000

7. If we use the OffsetsForTimes API,  the program could exit when timeout
    - ./kafkatest -broker=10.128.0.45:9093 -newAPI=false -topic=test -partitionId=2 -timeout=5000

conclusion:
I think the issue could be easily reproduced when a partitionId's leader(broker) is isolated.
The infinite looping code is [here](https://github.com/edenhill/librdkafka/blob/v0.11.6/src/rdkafka.c#L2643), 

**IMPORTANT**: Always try to reproduce the issue on the latest released version (see https://github.com/edenhill/librdkafka/releases), if it can't be reproduced on the latest version the issue has been fixed.

Checklist
=========

**IMPORTANT**: We will close issues where the checklist has not been completed.

Please provide the following information:

 - [x] librdkafka version (release number or git tag): `v0.11.6`
 - [x] Apache Kafka version: `confluent-oss-5.0.0-2.11`
 - [x] librdkafka client configuration: `"session.timeout.ms": 10000`
 - [x] Operating system: `centos 6`
 - [ ] Provide logs (with `debug=..` as necessary) from librdkafka
 - [ ] Provide broker log excerpts
 - [ ] Critical issue

### PR DESCRIPTION
**Pull request by milindl**
`rd_kafka_query_watermark_offsets` hangs forever if say, the broker is down, and exceeds the timeout in case the RTT > timeout, see the linked issue.

This PR adds a failing test, and then fixes the issue.

This also changes the fact the return value of `rd_kafka_q_serve` was being compared to RD_KAFKA_OP_RES_YIELD, which doesn't make sense to me, since it returns the count of served ops, and RD_KAFKA_OP_RES_YIELD is a positive constant.

Fixes #2588.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -34,6 +34,7 @@ librdkafka v2.3.0 is a feature release:
  * Fix to ensure max.poll.interval.ms is reset when rd_kafka_poll is called with
    consume_cb (#4431).
  * Fix for idempotent producer fatal errors, triggered after a possibly persisted message state (#4438).
+ * Fix `rd_kafka_query_watermark_offsets` hanging forever (#4460).
```
Comment (emasab) [inline CHANGELOG.md:37]:
```suggestion
 * Fix `rd_kafka_query_watermark_offsets` continuing beyond timeout expiry (#4460).
```

**Review by emasab — CHANGES_REQUESTED** [review]
Thanks for the fix! Just some test changes to ensure it's run when the test run has the local flag

File: src/rdkafka.c [hunk 2]
```diff
@@ -3590,11 +3591,15 @@ rd_kafka_resp_err_t rd_kafka_query_watermark_offsets(rd_kafka_t *rk,
...
-                                rd_kafka_poll_cb,
-                                NULL) != RD_KAFKA_OP_RES_YIELD)
-                ;
+        while (state.err == RD_KAFKA_RESP_ERR__IN_PROGRESS) {
+                tmout = rd_timeout_remains(ts_end);
+                if (rd_timeout_expired(tmout)) {
+                        state.err = RD_KAFKA_RESP_ERR__TIMED_OUT;
+                        break;
+                }
+                rd_kafka_q_serve(rkq, tmout, 0, RD_KAFKA_Q_CB_CALLBACK,
```
[the remark this answers is not in the document]
Comment (milindl) [reply src/rdkafka.c:3600]: Thanks!

File: tests/0031-get_offsets.c [hunk 3]
```diff
@@ -50,6 +49,11 @@ int main_0031_get_offsets(int argc, char **argv) {
         test_timing_t t_qry, t_get;
         uint64_t testid;
 
+        if (test_quick)
```
[the remark this answers is not in the document]
Comment (milindl) [reply tests/0031-get_offsets.c]: Done, the older test is now untouched and a new main is added.

**Comment by milindl** [comment]
Addressed comments, thanks for review @emasab

**Review by emasab — APPROVED** [review]
Thanks @milindl !

**Pull request merged**